### DISCUSSION TITLE
E3NN C++ version?

### DISCUSSION THREAD

**Discussion by ipcamit** [e3nn/e3nn#454] [Ideas]
Is there anyone working on a e3nn cpp version? I feel it might be useful to the community. If not, I am trying to it [here](https://github.com/ipcamit/e3nn-cpp), anyone interested is more than welcome to join! Idea is the following
```
Step 1: map all the python functions (torch version, as I found jax version bit harder to read though they might give better mileage in long run, see the Enzyme point) as closely as possible to original implementation
Step 2: Test every function simultaneously to ensure it gives expected output
Step 3: Use enzyme to generate the gradients, and perhaps a libe3nn.so etc.
```

Step 3 will be bit of a challenge given the experimental nature of Enzyme and the fact that it struggles with complicated C++ stl containers sometimes, but nevertheless will cross that bridge when it comes to that. Enzyme has better chances of being successful in functional code, like that of JAX, but I found the torch version much better organized to follow. If needed I will tune it later at Enzyme compilation stage.

I am just planning on chipping away one function at a time, every day or two. Any suggestions and contributions welcome! Currently working on `e3nn.math._reduce.py`

**Comment by mitkotak** [comment]
I think you should be able to borrow a lot of stuff from https://github.com/teddykoker/e3nn.c

**Comment by mitkotak** [comment]
Do you know if Enzyme be would able to backpropogate through torch.export generate C++ code cause I am close to getting an inference only C++ pipeline setup using torch.export.

**Comment by ipcamit** [comment]
Thanks for the suggestion I was not aware of `e3nn.c` It did not show up when I googled before starting. As far I understand, `torch.export` generates custom code, like torchscript graph right? Or are you getting C++ code out of `torch.export`? If it is a C/C++ code, I am sure Enzyme can handle that. If your C++ port is publicly accessible, I can volunteer some efforts to compile!

**Reply by mitkotak** [reply]
Yup it's C++ code which calls custom Triton kernels. Looks something like [this](https://gist.github.com/mitkotak/fa4aa2b247ed6597dd657af664075c4a). Here's how to call it for [TP](https://github.com/teddykoker/e3nn.c/pull/6/files). Let's do a POC for FullTensorProductv2 first.

**Reply by mitkotak** [reply]
The next step then can be to see if we can differentiate through a Nequip-Tetris that I am working on [here](https://github.com/atomicarchitects/e3nn_simple_network). Excited to see what happens. Big fan of Enzyme and Billy's work !

**Reply by ipcamit** [reply]
Triton Kernels might be trouble. Enzyme can differentiate through pure cuda and C++ right now. I know people are currently trying to go though Khronos library (it is not working yet). I think better will be to try some very simple functions first to check if it can handle the triton kernels etc yet.

`e3nn.c` should be differentiable through Enzyme. I can also try that first.

**Reply by mitkotak** [reply]
This might be straight up evil but Torch also added an experimental Halide backend that generates CUDA. It's extremely buggy though.
